lib/ansible/modules/cloud/misc/qemu_img.py

```
@@ -0,0 +1,212 @@
+  sample: "success"
+  returned: success
+'''
+
+import os
+
+from ansible.module_utils.basic import AnsibleModule
+from ansible.module_utils.pycompat24 import get_exception
+
+def main():
+
+    module = AnsibleModule(
+        argument_spec = dict(
+            dest = dict(type='str', required=True),
+            options = dict(type='str', default='preallocation=metadata'),
+            format = dict(type='str', default='qcow2'),
+            size = dict(type='str'),
+            grow= dict(type="bool", default=True),
+            shrink = dict(type="bool", default=False),
+            state = dict(type='str', choices=['absent', 'present'], default='present'),
+        ),
+        supports_check_mode = True,
+    )
+
+    result = dict(
+        changed = False,
+    )
+
+    state = module.params['state']
+    dest = module.params['dest']
+    img_format = module.params['format']
+    img_options = module.params['options']
+    size = module.params['size']
+    grow = module.params['grow']
+    shrink = module.params['shrink']
+
+    qemu_img = module.get_bin_path('qemu-img', True)
+
+    if state == 'present':
+        if not size:
+            module.fail_json(msg="No size defined, creating a disk image requires a size")
+        if not os.path.exists(dest):
+            if not module.check_mode:
+                if not img_options:
+                    rc, stdout, stderr = module.run_command('%s create -f %s "%s" %s' % (qemu_img, img_format, dest, size), check_rc=True)
+                else:
+                    rc, stdout, stderr = module.run_command('%s create -f %s -o "%s" "%s" %s' % (qemu_img, img_format, img_options, dest, size), check_rc=True)
+            result['changed'] = True
+        else:
+            try:
+                size_unit = str(size[-1:])
+                units = list('b', 'k', 'M', 'G', 'T', 'P', 'E', 'Z', 'Y')
+                unitmultiplier = 1024**units.index(size_unit)
+            except:
+                unitmultiplier = 1
```

This means that if you'd be using 16m, it will be 16 bytes too.
You will have to do a lot more to avoid mixups here.

---

Yes, it needs to be 16M to be MB, the only place this could be confusing is with k and K, because actually both should work, however only k will work.
I do not see this as a mix up, case of the unit size is sensitive.

---

The mixup to me is that people using incorrect units will silently be using bytes, whereas it should report this as an error.

---

That makes sense, we need to throw an error if the unit is now in the list and not presume that it is bytes